### PR TITLE
feat(#154): auto-refresh jobs-manager image on Docker Hub publish

### PR DESCRIPTION
**Pull request by saadqbal**
## Summary

Adds an `imageRefresh` CronJob that polls Docker Hub for new manifest digests of `tracebloc/jobs-manager` and `tracebloc/pods-monitor` under the floating `CLIENT_ENV` tag and runs `kubectl rollout restart` on the jobs-manager Deployment only when at least one digest has changed. Closes the gap left by `imagePullPolicy: Always` — that policy only pulls on pod creation, so running pods stay on the old image until something restarts them. Customers should not have to run any commands when we publish a new image.

Closes #154.

## Design

- Mirrors the existing `autoUpgrade` pattern (ConfigMap + CronJob + RBAC) so operators have one consistent mental model for the chart's background loops.
- **Narrower trust boundary than autoUpgrade**: namespace-scoped Role with `get,list pods` and `get,patch deployments` only — no cluster-admin. A compromise of this Pod is bounded to kicking the jobs-manager Deployment in the release namespace.
- **Idle-cheap**: each tick HEADs two manifests on Docker Hub; when both digests match the running pod, the script exits without touching the deployment. Default schedule (every 15m at :07/:22/:37/:52) stays well under Docker Hub's 100/6h anonymous pull-rate limit.
- **Per-image opt-out**: when `images.jobsManager.digest` or `images.podsMonitor.digest` is set, that image is skipped at runtime via env flags — digest pinning is an explicit reproducibility contract that auto-restart must not fight. If BOTH are pinned, the chart renders no CronJob, Role, RoleBinding, ConfigMap, or ServiceAccount at all.
- Manifest HEAD sends four `Accept` media types (Docker v2 + v2 list + OCI manifest + OCI index) so digest resolution works regardless of which variant the registry serves for that tag.
- Pod is PSA-restricted (runAsNonRoot, readOnlyRootFilesystem, dropped caps, RuntimeDefault seccomp); `HOME` points at a writable emptyDir so kubectl's discovery cache works under read-only root.

## Why not Keel?

Considered — Keel is the canonical "watch the registry, restart on new digest" controller. The benefits (semver policies, approval gates, multi-workload auto-discovery) don't pay off at two deployments on a mutable env tag with anonymous Docker Hub pulls. Revisit if we grow to ~5+ auto-refreshed workloads or need approval gates. See #154 discussion.

## Why not the ingestor?

Out of scope. Ingestor `hookImage` is `curlimages/curl` (third-party, pinned) and the workload is a Helm `post-install,post-upgrade` Job — `rollout restart` doesn't apply, and re-running a side-effecting hook on every image push would be wrong. Ingestor image refreshes ride along with chart upgrades via `autoUpgrade`.

## Test plan

- [x] `helm lint client/ --set clientId=test --set clientPassword=test`
- [x] `helm template` renders `stg-image-refresh` ServiceAccount, Role, RoleBinding, ConfigMap, and CronJob in the release namespace
- [x] `helm unittest client/` — 135/135 (15 new in `image_refresh_test.yaml`)
- [ ] Manual on a dev cluster: install the chart at this branch, confirm the CronJob runs on its first tick and exits with "all images up to date"
- [ ] Manual: push a new `tracebloc/jobs-manager:prod` image, wait for the next tick, confirm the deployment is rolled and pulls the new digest
- [ ] Manual: set `images.jobsManager.digest` to a pinned digest, helm upgrade, confirm the script logs "pinned by digest in values; skipping" and does not restart
- [ ] Manual: set BOTH `images.*.digest` values, helm upgrade, confirm none of the image-refresh resources render

## Chart version

Bumped 1.3.5 → 1.4.0 (feature add). `autoUpgrade` on already-deployed releases will pick this up on its next tick.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Adds a new background CronJob that polls Docker Hub and patches/restarts the `jobs-manager` Deployment, introducing operational risk around unexpected rollouts and reliance on external registry availability, though RBAC is kept namespace-scoped and limited to deployments.
> 
> **Overview**
> Bumps the Helm chart version to `1.4.0` and adds an **`imageRefresh` controller loop** that periodically checks Docker Hub manifest digests for `tracebloc/jobs-manager` and `tracebloc/pods-monitor` under the current `CLIENT_ENV` tag, then performs `kubectl rollout restart` (and records digests via deployment annotations) only when a digest changes.
> 
> This introduces new rendered resources (ConfigMap script, CronJob, ServiceAccount, namespace-scoped Role/RoleBinding) gated by `imageRefresh.enabled`, with additional logic to render nothing when *both* images are digest-pinned, plus new `values.yaml` defaults, `values.schema.json` validation (including rejecting `imageRefresh.image.tag: latest`), and Helm unit tests covering rendering/guardrails.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 877e11f37ec26cc28fa5809c37eb615a910d10b2. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: client/templates/image-refresh-rbac.yaml [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  # Read the running pod's container imageIDs to extract the active digest.
+  - apiGroups: [""]
+    resources: ["pods"]
+    verbs: ["get", "list"]
+  # `kubectl rollout restart` patches the deployment's pod template
+  # annotations (kubectl.kubernetes.io/restartedAt). `rollout status`
+  # reads the deployment.
+  - apiGroups: ["apps"]
+    resources: ["deployments"]
+    verbs: ["get", "patch"]
```
[the remark this answers is not in the document]
Comment (saadqbal) [reply client/templates/image-refresh-rbac.yaml]: Fixed in 88308c4. Added `watch` to the deployments verbs and tightened the RBAC test to lock the full verb set with `equal` (was `contains`) so a future refactor dropping `watch` fails loudly instead of silently regressing. 137/137 tests pass.

File: client/templates/image-refresh-cronjob.yaml [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+        return 1
+      fi
+      curl -fsSI --max-time 10 \
+        -H "Authorization: Bearer $_token" \
+        -H "Accept: application/vnd.docker.distribution.manifest.v2+json" \
+        -H "Accept: application/vnd.docker.distribution.manifest.list.v2+json" \
+        -H "Accept: application/vnd.oci.image.manifest.v1+json" \
+        -H "Accept: application/vnd.oci.image.index.v1+json" \
+        "https://registry-1.docker.io/v2/${_repo}/manifests/${IMAGE_TAG}" \
+        | awk 'BEGIN{IGNORECASE=1} /^docker-content-digest:/ {print $2}' \
```
[the remark this answers is not in the document]
Comment (saadqbal) [reply client/templates/image-refresh-cronjob.yaml]: Fixed in 834d005. Replaced `BEGIN{IGNORECASE=1}` with a `tr '[:upper:]' '[:lower:]'` case-fold pipe before awk — works on BusyBox awk (which alpine/k8s ships) regardless of HTTP/1.1 vs HTTP/2 header case. Digests are lowercase-by-spec so the captured value is unaffected. Added two regression tests: must include the `tr` pipe, must NOT include `BEGIN{IGNORECASE` in awk code. 137/137 tests pass.

File: client/templates/image-refresh-cronjob.yaml [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+      # HTTP/1.1 over a TLS-terminating proxy can return mixed case.
+      # alpine/k8s ships BusyBox awk, which silently ignores gawk's
+      # IGNORECASE=1, so we case-fold with tr BEFORE awk. Digests are
+      # lowercase-by-spec (sha256:<64 hex>), so tr never mangles the value.
+      curl -fsSI --max-time 10 \
+        -H "Authorization: Bearer $_token" \
+        -H "Accept: application/vnd.docker.distribution.manifest.v2+json" \
+        -H "Accept: application/vnd.docker.distribution.manifest.list.v2+json" \
+        -H "Accept: application/vnd.oci.image.manifest.v1+json" \
+        -H "Accept: application/vnd.oci.image.index.v1+json" \
```
[the remark this answers is not in the document]
Comment (saadqbal) [reply client/templates/image-refresh-cronjob.yaml]: Fixed in f384a99. Consolidated the four manifest-media-type values into a single comma-separated `Accept` header (spec-canonical form, avoids proxy-drop edge cases). Added a regression test matching the full joined header string so a future refactor splitting it back can't silently regress.

File: client/templates/image-refresh-rbac.yaml [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    verbs: ["get", "list"]
+  # `kubectl rollout restart` patches the deployment's pod template
+  # annotations (kubectl.kubernetes.io/restartedAt). `rollout status`
+  # uses a ListWatch on the deployment to wait for rollout completion;
+  # without `watch` it enters a reflector error loop (kubectl#580) and
+  # hangs until --timeout fires, marking every successful restart as a
+  # failed Job. Caught in PR #155 review.
+  - apiGroups: ["apps"]
+    resources: ["deployments"]
+    verbs: ["get", "patch", "watch"]
```
[the remark this answers is not in the document]
Comment (saadqbal) [reply client/templates/image-refresh-rbac.yaml]: Fixed in f384a99. Added `list` to the deployments verbs — symmetric failure mode to the missing `watch` (ListWatch needs both for the initial sync + incremental updates). Test now pins `["get", "list", "patch", "watch"]` with `equal` so dropping either fails loudly.

File: client/templates/image-refresh-cronjob.yaml [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+
+    # Read the running pod's imageID for a specific container by name.
+    # imageID format is `<registry>/<repo>@sha256:...`; we trim the prefix
+    # so it can be compared directly against the registry's digest.
+    get_running_digest() {
+      _container="$1"
+      kubectl get pod -n "$RELEASE_NAMESPACE" -l app=manager \
+        -o jsonpath="{.items[0].status.containerStatuses[?(@.name=='${_container}')].imageID}" \
+        | sed 's|.*@||' \
+        | tr -d '\r\n'
```
[the remark this answers is not in the document]
Comment (saadqbal) [reply client/templates/image-refresh-cronjob.yaml:100]: Fixed in cb3e180. When the top-level digest doesn't match running, the script now GETs the manifest body and, if it's an index (`"manifests"` key present), checks whether the running digest matches any per-platform manifest digest extracted via `grep -oE 'sha256:...'`. Only restarts if no match anywhere. Extra GET only fires on the rare mismatch path, so steady-state rate-limit usage is unchanged. Two regression tests added.

File: client/templates/image-refresh-cronjob.yaml [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+          break
+        fi
+      done
+
+      if [ "$_matched" -eq 1 ]; then
+        log "  running matches a per-platform manifest digest in the index; up to date"
+      else
+        log "  digest changed -> restart needed"
+        restart_needed=1
+      fi
```
[the remark this answers is not in the document]
Comment (saadqbal) [reply client/templates/image-refresh-cronjob.yaml:150]: Fixed in a937c75. `get_index_digests` now returns exit 1 on transient (token failure or non-2xx GET) vs exit 0 + empty stdout for the successfully-fetched single-platform case. Caller captures exit status separately via `if ! _index_digests=...` and `continue`s with a WARN log on transient — same skip-this-tick contract as the top-level HEAD path. Tests pin the capture pattern + skip log so the asymmetry can't creep back.

File: client/templates/image-refresh-cronjob.yaml [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+    # The media types are sent as a SINGLE comma-separated Accept header
+    # per the Docker registry v2 spec and the HTTP semantics RFC 7230 §3.2.2
+    # (which allows the syntactically-equivalent multi-line form, but the
+    # registry's documented form is the single-header one; some proxies
+    # have been observed dropping or reordering repeated Accept headers).
+    get_latest_digest() {
+      _repo="$1"
+      _token="$(get_token "$_repo")"
+      if [ -z "$_token" ]; then
+        log "  WARN: could not fetch auth token for $_repo"
```
[the remark this answers is not in the document]
Comment (saadqbal) [reply client/templates/image-refresh-cronjob.yaml:80]: Fixed in 877e11f as part of a broader refactor: `log()` now writes to stderr unconditionally (`printf ... >&2`), so command-substitution capture sites can't be polluted regardless. The broader refactor also rips out the imageID-vs-registry comparison entirely and switches to an annotation-based source of truth (`tracebloc.io/last-refreshed-<image>-digest` on the deployment), eliminating the root cause behind this and the previous four bugbot findings on this PR. Script shrank ~170→~115 lines, 136/136 tests pass.

**Review by aptracebloc — APPROVED** [review]
